redbot/setup.py

```diff
@@ -132,10 +132,11 @@ def get_name() -> str:
             " A-z, numbers, underscores, and hyphens."
         )
         name = input("> ")
-        if re.fullmatch(r"[a-zA-Z0-9_\-]*", name) is None:
+        if re.fullmatch(r"[a-zA-Z0-9_][a-zA-Z0-9_\-]*", name) is None:
```

Comment:
```suggestion
        if re.fullmatch(r"[a-zA-Z0-9]*", name) is None:
```

This is not final .... but like ... in all honesty ... this is an identifier ... with all the problem we seen due to symbols i would be leaning toward just making this identifier alpha numerical.

Reply:
I'd lean towards agreeing with this -- this would also need an owner DM/log warning for users who already have a `_ ` or `-` asking them to use `redbot --edit-instance` if we want users to transition away from this.

Reply:
I'm against it, I like my hyphens in instance names and also being able to use number at the beginning of the name, I would even want some other characters (like dots for example) allowed.

On the other hand, hyphens specifically (though there might be other characters that have this issue) need to be escaped when used in argument to systemd service. So while I would want my freedom, there are definitely some good reasons to forbid this.

What would you say about... Adding a prompt which will warn the user if the characters used in instance name *may* cause problems (i.e. hyphen in name)? On the bad side, this would need us to define 2 patterns (or give complete freedom which I'm not so sure is a good idea either, though arguably not terrible if it explicitly says it can cause problems) but on the bright side, it gives more freedom when choosing the instance name.

Reply:
Oh and I forgot, I don't think there's really a need to *transition* users to new restrictions. Until 3.2, Red allowed any character except spaces and we did not do it - it's not like old instance names don't work and if there were some issues with them, I think the user would already notice it quite early.

Reply:
I can get behind the prompt.

I mean if we allow alphanumetic `string.isalnum` could do the trick, then regex is used for prompt less strict. what do you think?

Reply:
I'd be happy with that, though I don't use hyphens in instance names, as well or we don't have hyphens and continue to allow underscores `_`.

That's two nos for a transition then! Fine by me.

So draper you mean that if it's not alphanumeric there's a prompt warning about possible issues?

Reply:
Yeah thats the way i'd go about it.

If not alpha-numeric prompt for warning about potential issues, then if they say yes validate their entry agaisnt the more lenient regex.

Reply:
> I mean if we allow alphanumetic `string.isalnum` could do the trick

`.isalnum()` is too lenient to be used here - it allows characters like "Ą". I think going with `[A-Za-z0-9_]+` regex (so just removing the hyphen from current one) would be the best to go about it for the strict one. As far as I know, there's nothing in Python that would make it easier than that, the closest is `str.isidentifier()` though I don't think there's a need to be *that* strict in the strict check :)

> If not alpha-numeric prompt for warning about potential issues, then if they say yes validate their entry agaisnt the more lenient regex.

Is there any reason not to validate the name against the more lenient regex *before* prompting? (only if it fails the strict one and we would actually need to prompt of course)

Now that I think about it... We could simply do a regex `[A-Za-z0-9_\.\-]+` and warn `if "-" in string` 🤔 I don't know, this doesn't really give much more freedom *and only adds what I want* while making it so regular user doesn't use the problematic hyphens... But I honestly don't know what character should be allowed in lenient check... Hyphen specifically proven to be problematic, so warning about it makes sense, but other than that I'm not sure if splitting into strict and lenient check makes as much sense as I initially thought it would. Maybe I should give up on hyphens, I don't know at this point :/

Reply:
Maybe I should add a disclaimer that I'm a windows user and that hyphens work great here 😄 (well, other than the case where you start instance name with hyphen)